lib/Lower/Bridge.cpp

```diff
@@ -1038,18 +1249,65 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         createTemporary(toLocation(), details->result());
       }
     }
+
+    // Create most function blocks in advance.
+    createEmptyBlocks(funit.evaluationList);
+
+    // Reinstate entry block as the current insertion point.
+    builder->setInsertionPointToEnd(&func.front());
+  }
+
+  /// Create empty blocks for the current function.
+  void createEmptyBlocks(
+      std::list<Fortran::lower::pft::Evaluation> &evaluationList) {
+    for (auto &eval : evaluationList) {
+      if (eval.isNewBlock) {
+        eval.block = Fortran::lower::createBlock(builder);
+      }
+      for (size_t i{0}; i < eval.localBlocks.size(); ++i) {
+        eval.localBlocks[i] = Fortran::lower::createBlock(builder);
+      }
+      if (eval.isConstruct()) {
+        if (eval.LowerAsUnstructured()) {
+          createEmptyBlocks(*eval.evaluationList);
+        } else {
+          // A structured construct that is a target starts a new block.
+          Fortran::lower::pft::Evaluation &constructStmt{
+              eval.evaluationList->front()};
+          if (constructStmt.isNewBlock) {
+            constructStmt.block = Fortran::lower::createBlock(builder);
+          }
+        }
+      }
+    }
+  }
+
+  /// Return the predicate: "current block does not have a terminator branch".
+  bool blockIsUnterminated() {
+    mlir::Block *currentBlock{builder->getBlock()};
+    return currentBlock->empty() || currentBlock->back().isKnownNonTerminator();
+  }
+
+  /// Unconditionally switch code insertion to a new block.
+  void startBlock(mlir::Block *newBlock) {
+    assert(newBlock && "missing block");
+    // If the current block does not have a terminator branch,
+    // append a fallthrough branch.
+    if (blockIsUnterminated()) {
+      genFIRUnconditionalBranch(newBlock);
+    }
+    builder->setInsertionPointToStart(newBlock);
   }
 
-  void finalizeQueuedEdges() {
-    for (auto &edgeFunc : localEdgeQ)
-      edgeFunc(localBlockMap);
-    localEdgeQ.clear();
-    localBlockMap.clear();
+  /// Conditionally switch code insertion to a new block.
+  void maybeStartBlock(mlir::Block *newBlock) {
+    if (newBlock) {
+      startBlock(newBlock);
+    }
```

Comment:
Statements that are branch targets must start a new block, but it needn't be necessary to check every time if a statement is a target or not.  Function maybeStartBlock allows that determination to made just once in a wrapper call for applicable cases, rather than placing a startBlock call under an IF for each such case.  At the moment there are 2 such cases.  Function startBlock can be called directly when the caller knows there **must** be a new block.  This division of concerns also allows adding an assert to just the cases where a block **must** be present.

Reply:
Ok.